velox/common/memory/Memory.h

```diff
@@ -144,24 +137,55 @@ class MemoryPool : public std::enable_shared_from_this<MemoryPool> {
       const std::string& name,
       int64_t cap = kMaxMemory);
 
-  /// Invoked to allocate a buffer with specified 'size'.
+  /// Allocates a buffer with specified 'size'.
   virtual void* FOLLY_NULLABLE allocate(int64_t size) = 0;
 
-  /// Invoked to allocate a zero-filled buffer with capacity that can store
-  /// 'numEntries' entries with each size of 'sizeEach'.
+  /// Allocates a zero-filled buffer with capacity that can store 'numEntries'
+  /// entries with each size of 'sizeEach'.
   virtual void* FOLLY_NULLABLE
   allocateZeroFilled(int64_t numEntries, int64_t sizeEach) = 0;
 
-  /// Invoked to re-allocate from an existing buffer with 'newSize' and update
-  /// memory usage counting accordingly. If 'newSize' is larger than the current
-  /// buffer 'size', the function will allocate a new buffer and free the old
-  /// buffer.
+  /// Re-allocates from an existing buffer with 'newSize' and update memory
+  /// usage counting accordingly. If 'newSize' is larger than the current buffer
+  /// 'size', the function will allocate a new buffer and free the old buffer.
   virtual void* FOLLY_NULLABLE
   reallocate(void* FOLLY_NULLABLE p, int64_t size, int64_t newSize) = 0;
 
-  /// Invoked to free an allocated buffer.
+  /// Frees an allocated buffer.
   virtual void free(void* FOLLY_NULLABLE p, int64_t size) = 0;
 
+  /// Allocates one or more runs that add up to at least 'numPages', with the
+  /// smallest run being at least 'minSizeClass' pages. 'minSizeClass' must be
+  /// <= the size of the largest size class. The new memory is returned in 'out'
+  /// and any memory formerly referenced by 'out' is freed. The function returns
+  /// true if the allocation succeeded. If returning false, 'out' references no
+  /// memory and any partially allocated memory is freed.
+  virtual bool allocateNonContiguous(
```

Comment:
Taking a look at MemoryAllocatorImpl::allocateNonContiguous() method and it unnaturally and artificially forces malloc to adapt to the size class system which was initially developed only to serve our own MmapAllocator purpose. I wonder why do we want to do that?

Reply:
Not sure if I understand your question. Let's chat offline about this. thanks!